drift/lib/src/runtime/query_builder/expressions/window.dart

```diff
@@ -0,0 +1,389 @@
+  /// );
+  ///
+  /// // Use in a query
+  /// final query = select(items).addColumns([runningTotal]);
+  /// ```
+  ///
+  /// # Partitioning
+  ///
+  /// Partition data into groups before applying the window function:
+  ///
+  /// ```dart
+  /// // Running total per category
+  /// final runningTotalByCategory = WindowFunctionExpression<int>(
+  ///   items.price.sum(),
+  ///   orderBy: [OrderingTerm.asc(items.date)],
+  ///   partitionBy: [items.categoryId],
+  /// );
+  /// ```
+  ///
+  /// # Frame Boundaries
+  ///
+  /// ## Rows Frame
+  /// Operates on physical rows:
+  ///
+  /// ```dart
+  /// // Moving average of previous 3 rows and current row
+  /// final movingAverage = WindowFunctionExpression<double>(
+  ///   items.price.avg(),
+  ///   orderBy: [OrderingTerm.asc(items.date)],
+  ///   boundary: RowsFrameBoundary(
+  ///     start: -3, // 3 rows preceding
+  ///     end: 0,    // current row
+  ///   ),
+  /// );
+  /// ```
+  ///
+  /// ## Groups Frame
+  /// Operates on groups of rows with same ORDER BY values:
+  ///
+  /// ```dart
+  /// // Include current group and one group before
+  /// final groupFrame = WindowFunctionExpression<int>(
+  ///   items.quantity.sum(),
+  ///   orderBy: [OrderingTerm.asc(items.category)],
+  ///   boundary: GroupsFrameBoundary(
+  ///     start: null, // Unbounded preceding groups
+  ///     end: 0,    // current group
+  ///   ),
+  /// );
+  /// ```
+  ///
+  /// ## Range Frame
+  /// Operates on value ranges (default):
+  ///
+  /// ```dart
+  /// // Sum of items within price range of ±10 from current row
+  /// final rangeFrame = WindowFunctionExpression<int>(
+  ///   items.quantity.sum(),
+  ///   orderBy: [OrderingTerm.asc(items.price)],
+  ///   boundary: RangeFrameBoundary(
+  ///     start: -10, // 10 units less than current value
+  ///     end: 10,    // 10 units more than current value
+  ///   ),
+  /// );
+  /// ```
+  ///
+  WindowFunctionExpression(
+    this.function, {
+    required this.orderBy,
+    this.partitionBy,
+    this.boundary = const RangeFrameBoundary(),
+  }) {
+    if (orderBy.isEmpty) {
+      throw ArgumentError.value(
+        orderBy,
+        'orderBy',
+        'Must not be empty',
+      );
+    }
+  }
+
+  @override
+  void writeInto(GenerationContext context) {
+    function.writeInto(context);
+    context.buffer.write(' OVER (');
+    if (partitionBy case final partitionBy? when partitionBy.isNotEmpty) {
+      PartitionBy(partitionBy).writeInto(context);
+      context.writeWhitespace();
+    }
+    OrderBy(orderBy).writeInto(context);
+    context.writeWhitespace();
+    boundary.writeInto(context);
+    context.buffer.write(')');
+  }
+}
+
+/// A partition-by clause as part of a window function statement. The clause can consist
+/// of multiple [Expression]s, with the first terms being primary partition and
+/// the later terms will work as a nested partition for the previous partition and so on.
+class PartitionBy extends Component {
```

Comment:
Since this class is only used to implement `writeInto`, let's make it private.

Reply:
Done [e91953a](https://github.com/simolus3/drift/pull/3474/commits/e91953a15483c508dff97f99c03c83d7b7cac2ea)